workspace_host/interface.js

```diff
@@ -335,17 +336,48 @@ function _recursiveUploadJobManager(curDirPath, S3curDirPath) {
     return ret;
 }
 
+// Extracts `workspace_id` and `/path/to/file` from `/prefix/workspace-${uuid}/path/to/file`
+function _getWorkspaceByPath(path) {
+    let localPath = path.replace(`${workspacePrefix}/`, '').split('/');
+    const localName = localPath.shift();
+    localPath = localPath.join('/');
+
+    if (typeof id_workspace_mapper === 'undefined') {
+        logger.error(`_getWorkspaceByLocalPath() error: id_workspace_mapper undefined`);
+        return;
+    }
+    const workspace_id = Object.keys(id_workspace_mapper).find(
+        key => id_workspace_mapper[key].localName === localName,
+    );
+    
+    return {workspace_id, localPath};
+}
+
 function _autoUpdateJobManager() {
     var jobs = [];
     console.log(`watch update_queue: ${JSON.stringify(update_queue)}`);
     for (const path in update_queue) {
+        const {workspace_id, localPath} = _getWorkspaceByPath(path);
+        logger.info(`watch: workspace_id=${workspace_id}, localPath=${localPath}`);
+        let s3Path;
+        if (!workspace_id) {
+            logger.info(`watch return: workspace_id not mapped yet`);
+            return;
+        } else if (localPath === '') {
+            logger.info(`watch continue: empty (root) path`);
+            continue;
+        } else {
+            s3Path = `workspace-${workspace_id}/${localPath}`;
```

Comment:
Should this be using `s3Name`?

Reply:
I think there's a race condition where fswatcher instantly sees the dirs but `s3Name` might not have been assigned yet by the workspace initializer.

Reply:
As we discussed in the meeting, I think we should be checking that the `id_workspace_mapper` is correctly initialized for this workspace and returning an error if it isn't. An error here probably means there is an old leftover workspace still running that we should be ignoring.

Reply:
Handled in `_getWorkspaceByPath()` now:

https://github.com/PrairieLearn/PrairieLearn/pull/2777/files#diff-e7517976b89db12af19f23b917645289R422-R427